lib/default.nix

```diff
@@ -90,33 +90,57 @@
     postPatch ? "", ...
   }: let
     vendoredSourceJson = "${dirOf pkg.meta.position}/vendored-source.json";
-    inherit (builtins) stringLength substring pathExists;
+    inherit (builtins) stringLength substring pathExists length;
     nameStart = 5 + stringLength pkg.rosDistro; # e.g. ros-jazzy- => 10
     attr = substring nameStart (-1) pkg.pname;
     errMsg = ''
       error: File ${vendoredSourceJson} missing.
       Run "$(nix-build -A rosPackages.${pkg.rosDistro}.${attr}.updateAmentVendor)" to create it.
     '';
-    sourceInfo = builtins.fromJSON (builtins.readFile vendoredSourceJson);
+    sourceInfos = builtins.fromJSON (builtins.readFile vendoredSourceJson);
     # ament_vendor doesn't allow patches for path inputs, so we have to pack it
     # into a tar first. Additionally, vcstool only accepts tarballs with the
     # version number as the root directory name.
-    vendor = lib.tarSource tarSourceArgs (
+    vendor = sourceInfo: lib.tarSource tarSourceArgs (
       self.fetchgit (sourceInfo // fetchgitArgs // {
         name = sourceInfo.rev;
       }));
   in {
     # CMake ExternalProject patches are applied with git apply
     nativeBuildInputs = nativeBuildInputs ++ [ self.git ];
-    postPatch = (if pathExists vendoredSourceJson then ''
-      sed -i '\|VCS_URL\s*|c\
-        VCS_URL "file://${vendor}"\
-        VCS_TYPE tar' \
-        ${lib.escapeShellArg file}
-    '' else ''
-      echo >&2 ${lib.escapeShellArg errMsg}
-      exit 1
-    '') + postPatch;
+    postPatch = (
+      if pathExists vendoredSourceJson then
+        # Remove all existing VCS_TYPE lines
+        ''
+          sed -i '/VCS_TYPE \(git\|zip\|svn\|path\)/d' ${lib.escapeShellArg file}
+        '' + (
+          if length sourceInfos == 1 then
```

Comment:
This seems a bit brittle. Not that big a deal since we only have to handle a limited number of examples, but as an alternative, would it be feasible to match against the original `VCS_URL` value when patching?

Reply:
Unfortunately not. CMakeFile authors are very creative and use variables in URLs like:

    VCS_URL https://github.com/gazebosim/${GITHUB_NAME}.git

I was also looking at whether `cmake` can tell us the line number from where a macro (of a function) was called so that we can match based on that, but nothing like this seems to exist.

Reply:
I guess it could be done if there were some way to wrap the `ament_vendor()` call with our own macro that replaces the `VCS_URL` parameter.

Reply:
I think the problem to solve here would be how to call the original `ament_vendor` implementation from the wrapper with the same name. I don't see any means for macro namespacing in cmake. Maybe, we can use Nix to patch the original implementation to have a different name and call that from the wrapper.